bentoml/_internal/utils/formparser.py

```diff
@@ -121,23 +131,48 @@ async def parse(self) -> _ItemsBody:
                 elif message_type == MultiPartMessage.HEADER_END:  # type: ignore
                     field = header_field.lower()
                     if field == b"content-disposition":
-                        _, options = multipart.parse_options_header(header_value)
-                        options = t.cast(t.Dict[bytes, bytes], options)
-                        field_name = user_safe_decode(options[b"name"], charset)
+                        content_disposition = header_value
+                    elif field == b"content-type":
+                        content_type = header_value
                     elif field == b"bentoml-payload-field":
                         field_name = user_safe_decode(header_value, charset)
                     else:
                         headers.append((field, header_value))
                     header_field = b""
                     header_value = b""
                 elif message_type == MultiPartMessage.HEADERS_FINISHED:  # type: ignore
-                    assert (
-                        field_name
-                    ), "`Content-Disposition` is not available in headers"
+                    if content_disposition is None:
+                        raise BadInput(
+                            'The Content-Disposition header field "name" must be provided.'
+                        )
+                    _, options = multipart.parse_options_header(content_disposition)
+                    options = t.cast(t.Dict[bytes, bytes], options)
+                    try:
+                        field_name = user_safe_decode(options[b"name"], charset)
+                    except KeyError:
+                        raise BadInput(
+                            'The Content-Disposition header field "name" must be provided.'
+                        )
+                    if b"filename" in options:
+                        filename = user_safe_decode(options[b"filename"], charset)
+                        multipart_file = UploadFile(
+                            filename=filename,
+                            content_type=content_type.decode("latin-1"),
+                            headers=Headers(raw=headers),  # type: ignore (incomplete starlette types)
+                        )
+                    else:
+                        multipart_file = None
```

Comment:
Is it a valid format if "filename" is not a part of the header? If yes, can we still create an `UploadFile` to buffer the bytes?

Reply:
I think technically this is the case, but given that I think this will probably slow down small requests (this is entirely a hunch based on no evidence) I'm not sure we should do that.

FWIW this is starlette's current implementation of multipart, so at the very least we should match its performance now. I think @larme's cpython rewrite should be the real solution, though.